src/vs/editor/contrib/documentSymbols/outlineModel.ts

```diff
@@ -190,14 +193,14 @@ export class OutlineGroup extends TreeElement {
 		}
 
 		let myMarkers: IMarker[] = [];
-		let myTopSev: MarkerSeverity;
+		let myTopSev: MarkerSeverity | undefined = undefined;
 
 		for (; start < markers.length && Range.areIntersecting(item.symbol.range, markers[start]); start++) {
 			// remove markers intersecting with this outline element
 			// and store them in a 'private' array.
 			let marker = markers[start];
 			myMarkers.push(marker);
-			markers[start] = undefined;
+			filteredMarkers[start] = undefined;
```

Comment:
it's important to modify `markers`, don't create another another "name" for it

Reply:
The issue is that `markers` is of type `Array<IMarker>` which can't accept undefined values. By the end of the function, those undefined values are removed through `coalesceInPlace`, but in the meantime we need to let the type system know about the temporary undefineds.

I tried changing the type of `markers` to `Array<IMarker | undefined>` instead, but then more errors get throw for uses of `binarySearch`, `Range.areIntersecting` and others in that function so I went for a new alias as the solution that requires minimal code changes, though I agree it feels like a pretty hacky solution. Alternatives could include rewriting some of the logic to be a little it more functional (e.g. use filter) so that the array doesn't contain temporary undefined values, but that requires even more logic changes.

Reply:
when in doubt always use the `!` operator. that makes it a lot easier for me/us to reason about these changes.

Reply:
Right that makes perfect sense, I'm just not sure how to apply it in this case because the `!` can't operate in a templated type (turn `Array<T | undefined>` into `Array<T>`). In this case, `markers` is actually both `Array<T | undefined>` _and_  `Array<T>` at different points in the function due to mutation.

What do you think of this other idea? Use type casting: `(markers as Array<IMarker | undefined>)[start] = undefined;`